crates/common/src/state_update.rs

```diff
@@ -208,6 +210,186 @@ impl StateUpdate {
+            .finish()
+    }
+
+    fn hash_of_declared_classes(&self) -> FieldElement {
+        let declared_classes: BTreeSet<(SierraHash, CasmHash)> = self
+            .declared_sierra_classes
+            .iter()
+            .map(|(sierra, casm)| (*sierra, *casm))
+            .collect();
+
+        let number_of_declared_classes = declared_classes.len() as u64;
+
+        declared_classes
+            .iter()
+            .fold(
+                {
+                    let mut hasher = PoseidonHasher::new();
+                    hasher.write(number_of_declared_classes.into());
+                    hasher
+                },
+                |mut hasher, (sierra, casm)| {
+                    hasher.write(sierra.0.into());
+                    hasher.write(casm.0.into());
+                    hasher
+                },
+            )
+            .finish()
+    }
+
+    fn hash_of_old_declared_classes(&self) -> FieldElement {
+        let declared_classes: BTreeSet<ClassHash> =
+            self.declared_cairo_classes.iter().copied().collect();
+
+        let number_of_declared_classes = declared_classes.len() as u64;
+
+        declared_classes
+            .iter()
+            .fold(
+                {
+                    let mut hasher = PoseidonHasher::new();
+                    hasher.write(number_of_declared_classes.into());
+                    hasher
+                },
+                |mut hasher, class_hash| {
+                    hasher.write(class_hash.0.into());
+                    hasher
+                },
+            )
+            .finish()
+    }
+
+    fn hash_of_storage_domain_state_diff(&self) -> FieldElement {
+        let storage_diffs = self
+            .contract_updates
+            .iter()
+            .filter_map(|(address, update)| {
+                if update.storage.is_empty() {
+                    None
+                } else {
+                    let updates = update
+                        .storage
+                        .iter()
+                        .map(|(key, value)| (*key, *value))
+                        .collect();
+
+                    Some((*address, updates))
+                }
+            });
+        let system_storage_diffs =
+            self.system_contract_updates
+                .iter()
+                .filter_map(|(address, update)| {
+                    if update.storage.is_empty() {
+                        None
+                    } else {
+                        let updates: BTreeMap<StorageAddress, StorageValue> = update
+                            .storage
+                            .iter()
+                            .map(|(key, value)| (*key, *value))
+                            .collect();
+
+                        Some((*address, updates))
+                    }
+                });
+        let storage_diffs: BTreeMap<ContractAddress, BTreeMap<StorageAddress, StorageValue>> =
+            storage_diffs.chain(system_storage_diffs).collect();
```

Comment:
Are the `BTreeMap` collections used to sort the iteration? Or what is the purpose?

Reply:
Yes. According to the specification the "flattened" representation needs to be sorted.

Reply:
Took me a minute - thought you were collecting just to count 😁